source/slang/slang-ir-lower-generics.cpp

```diff
@@ -208,6 +237,133 @@ namespace Slang
+            }
+            IRBuilder builderStorage;
+            auto builder = &builderStorage;
+            builder->sharedBuilder = &sharedBuilderStorage;
+            builder->setInsertBefore(storeInst);
+            auto size = builder->emitRTTIExtractSize(rttiType->getRTTIOperand());
+            auto copy = builder->emitCopy(storeInst->ptr.get(), valPtr, size);
+            storeInst->replaceUsesWith(copy);
+            storeInst->removeAndDeallocate();
+        }
+
+        void processLoadInst(IRLoad* loadInst)
+        {
+            auto rttiType = as<IRRTTIPointerType>(loadInst->ptr.get()->getFullType());
+            if (!rttiType)
+                return;
+            // All loads of generic typed variables must be eliminated.
+            // There are only two possible uses of a load(genericVar):
+            // 1. store(x, load(genVar)), which will be handled by processStoreInst.
+            // 2. call(f, load(genVar)) when calling a generic function or a member function
+            //    via an interface witness lookup. In this case, we need to replace with
+            //    just `genVar`, since that function has already been lowered to take
+            //    raw pointers.
+            // Here we replace all uses of load to just the pointer itself.
+            // After this, all arguments in `call`s will be in its correct form.
+            // All `store`s will become `store(x, genVar)`, and still need
+            // to be translated into a `copy`, we leave that step when we get to
+            // process the `store` inst.
+            loadInst->replaceUsesWith(loadInst->getOperand(0));
+            loadInst->removeAndDeallocate();
+        }
+
+        // Emits an IRRTTIObject containing type information for a given type.
+        IRInst* maybeEmitRTTIObject(IRInst* typeInst)
+        {
+            IRInst* result = nullptr;
+            if (rttiObjects.TryGetValue(typeInst, result))
+                return result;
+            IRBuilder builderStorage;
+            auto builder = &builderStorage;
+            builder->sharedBuilder = &sharedBuilderStorage;
+            builder->setInsertBefore(typeInst);
+
+            // For now the only type info we encapsualte is type size.
+            IRSizeAndAlignment sizeAndAlignment;
+            getNaturalSizeAndAlignment((IRType*)typeInst, &sizeAndAlignment);
+            auto typeSize = builder->getIntValue(builder->getIntType(), sizeAndAlignment.size);
+            auto sizeEntry = builder->emitRTTIEntry(kRTTISize, typeSize);
+            result = builder->emitMakeRTTIObject(makeArrayView(sizeEntry));
+
+            // Give a name to the rtti object.
+            if (auto exportDecoration = typeInst->findDecoration<IRExportDecoration>())
+            {
+                String rttiObjName = "__rtti_" + String(exportDecoration->getMangledName());
+                builder->addExportDecoration(result, rttiObjName.getUnownedSlice());
+            }
```

Comment:
Can the RTTI object just use the mangled name of the type itself for export? At this point we are late enough in the compiler that we are concerned with the linker of the target machine and not the Slang linker, so there is no need to worry about clashes with the mangled name of the type itself. Prefixing the name is fine for now, but seems like overkill.

(Also: one minor note is that some platforms/toolchains apparently support compressing string tables for binaries based on the idea that many symbol names will have long prefixes in common. As a result, we should tend to pick mangling strategies that make related symbols have the same prefix, which means adding the text that differentiates symbols to the end of the name instead of the beginning.)

Reply:
Right now we don't use the mangled names for symbols *at all* when emitting code, so the mangled names don't really affect anything downstream from the IR linker.

Where this would start to be an issue is if we wanted to compile `a.slang` -> `a.cpp` and then `b.slang` -> `b.cpp` and then compile and link both `a.cpp` and `b.cpp` together (resolving cross-references between them), which seems to be exactly where all your current work is headed (in terms of capabilities).

For `a.cpp` to refer to a function that will be deifned in `b.cpp` they will need a way to agree on the name that will be used for the platform linker, and the current strategy of generating unique names by taking the "name hint" decoration and adding a numeric suffix obviously won't work (one module could name something `foo_1` and the other could name the same thing `foo_3` because there happened to be other things named `foo`).

The right fix would be that when we are in "separate compilation mode" (or however we want to enable this stuff), we emit all the symbols in our generated C/C++ as `extern "C"` symbols using the Slang-generated mangled name:

```c++
// a.cpp
extern "C" int _S1b3foo...(float x);

// b.cpp
extern "C" int _S1b3foo...(float x)
{
    ...
}
```

All that matters for making the platform linker do the right thing is that modules agree on the mangled name for symbols that have linkage. Because we are using `extern "C"` the name is all that matters, and the names of the types of parameters/results don't matter.

The point I was getting at is then that *types* do not have linkage in terms of traditional platform linkers; there is no such thing as an `extern "C" struct` in C/C++. As such, we would never need to emit a `struct Bar` using a mangled name in generated C/C++ code, and there would be no problem with using `Bar_0` in one generated file and `Bar_1` in another (again, because we use `extern "C"` linkage for all functions, so that the names of parameter types don't affect the name of the function's symbol).

Because the type itself doesn't need to be emitted using a mangled name, the corresponding mangled name is available and free to use for any `extern "C"` RTTI object declarations/definitions we want to emit.